ddtrace/tracer/civisibility_transport.go

```diff
@@ -0,0 +1,201 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+package tracer
+
+import (
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/constants"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
+)
+
+// Constants for CI Visibility API paths and subdomains.
+const (
+	TestCycleSubdomain = "citestcycle-intake" // Subdomain for test cycle intake.
+	TestCyclePath      = "api/v2/citestcycle" // API path for test cycle.
+	EvpProxyPath       = "evp_proxy/v2"       // Path for EVP proxy.
+)
+
+// Ensure that civisibilityTransport implements the transport interface.
+var _ transport = (*ciVisibilityTransport)(nil)
+
+// ciVisibilityTransport is a structure that handles sending CI Visibility payloads
+// to the Datadog endpoint, either in agentless mode or through the EVP proxy.
+type ciVisibilityTransport struct {
+	config           *config           // Configuration for the tracer.
+	testCycleURLPath string            // URL path for the test cycle endpoint.
+	headers          map[string]string // HTTP headers to be included in the requests.
+	agentless        bool              // Gets if the transport is configured in agentless mode (eg: Gzip support)
+}
+
+// newCiVisibilityTransport creates and initializes a new civisibilityTransport
+// based on the provided tracer configuration. It sets up the appropriate headers
+// and determines the URL path based on whether agentless mode is enabled.
+//
+// Parameters:
+//
+//	config - The tracer configuration.
+//
+// Returns:
+//
+//	A pointer to an initialized civisibilityTransport instance.
+func newCiVisibilityTransport(config *config) *ciVisibilityTransport {
+	// Initialize the default headers with encoder metadata.
+	defaultHeaders := map[string]string{
+		"Datadog-Meta-Lang":             "go",
+		"Datadog-Meta-Lang-Version":     strings.TrimPrefix(runtime.Version(), "go"),
+		"Datadog-Meta-Lang-Interpreter": runtime.Compiler + "-" + runtime.GOARCH + "-" + runtime.GOOS,
+		"Datadog-Meta-Tracer-Version":   version.Tag,
+		"Content-Type":                  "application/msgpack",
+	}
+	if cid := internal.ContainerID(); cid != "" {
+		defaultHeaders["Datadog-Container-ID"] = cid
+	}
+	if eid := internal.EntityID(); eid != "" {
+		defaultHeaders["Datadog-Entity-ID"] = eid
+	}
+
+	// Determine if agentless mode is enabled through an environment variable.
+	agentlessEnabled := internal.BoolEnv(constants.CiVisibilityAgentlessEnabledEnvironmentVariable, false)
+
+	testCycleURL := ""
+	if agentlessEnabled {
+		// Agentless mode is enabled.
+		APIKeyValue := os.Getenv(constants.APIKeyEnvironmentVariable)
+		if APIKeyValue == "" {
+			log.Error("An API key is required for agentless mode. Use the DD_API_KEY env variable to set it")
+			os.Exit(1)
```

Comment:
Not sure about os.Exit, should there be a more graceful shutdown? I didn't find such exits in the existing code

Reply:
@dianashevchenko what is the recommended way to exit the tracer when the setup is wrong (like in this case there is no API key and we cannot send events to the intake)?

Reply:
I've removed `os.Exit`, now the app doesn't exit. In case the API key is not specified, the agent will send events to the intake but the requests will fail with 400.
As discussed offline, CI visibility library will be internal until the tracer v2 implementation is finished, and in the v2 we'll be able to return an error in such cases.
For now, I'm going to set up a monitor for this log message to track if it happens.